### ISSUE TITLE
Wrong spans for loop invariants

### ISSUE THREAD

**Issue by Armael** [creusot-rs/creusot#1225]
On the following repro example (not minimal) (needs #1224), after splitting on `vc_knuth_shuffle`, the spans of many of the loop invariants (lines 128-136) are incorrect. For instance, subgoal 43 corresponds to the last invariant, but its span points to the first invariant.

```rust
use creusot_contracts::*;

// Axiomatized "Rationals"

pub enum Q {
    Zero,
    Frac1(Int),
}
use Q::*;

impl Q {
    #[trusted]
    #[logic]
    #[ensures(self == Zero ==> result == Zero)]
    #[ensures(forall<m:Int> self == Frac1(n * m) ==>
        result == Frac1(m))]
    fn mul_int(self, n: Int) -> Q {
        absurd
    }
}

// Axiomatized Error credits + corresponding "rand_in_range()" function

pub struct ErrCredit {
    amount: Q,
}

impl ErrCredit {
    #[trusted]
    #[logic]
    #[requires(self.amount == Frac1(1))]
    #[ensures(false)]
    fn impossible(self) -> bool {
        absurd
    }
}

// NOTE: does not require `avoid` to be in range!
#[trusted]
#[requires(l@ <= u@)]
#[ensures(l@ <= result.0@ && result.0@ < u@)]
#[ensures(
    if *result.1 {
        result.0@ != *avoid &&
        (^e.inner_logic()).amount == Zero
    } else {
        result.0@ == *avoid &&
        (^e.inner_logic()).amount == e.amount.mul_int(u@ - l@)
    }
)]
fn rand_in_range(
    l: usize,
    u: usize,
    e: GhostBox<&mut ErrCredit>,
    avoid: Snapshot<Int>,
) -> (usize, Snapshot<bool>) {
    panic!()
}

// factorial

#[logic]
#[requires(n >= 0)]
#[variant(n)]
fn fact(n: Int) -> Int {
    if n <= 0 {
        1
    } else {
        n * fact(n-1)
    }
}

// Properties of sequences: nodup, subsequences

#[predicate]
fn inbounds<T>(s: Seq<T>, i: Int) -> bool {
    0 <= i && i < s.len()
}

#[logic]
#[trusted]
#[requires(s.contains(x))]
#[ensures(inbounds(s, result) && s[result] == x)]
fn find_idx<T: Sized>(s: Seq<T>, x: T) -> Int {
  absurd // TODO: actual proof? general existential elimination principle?
}

#[predicate]
fn nodup<T>(s: Seq<T>) -> bool { pearlite! {
    forall<i: Int, j: Int> inbounds(s, i) && inbounds(s, j) && i != j ==>
        s[i] != s[j]
}}

#[logic]
#[requires(nodup(s1))]
#[requires(s2.permutation_of(s1))]
#[ensures(nodup(s2))]
fn permut_nodup<T>(s1: Seq<T>, s2: Seq<T>) {
    // TODO
}

#[logic]
#[ensures(forall<s0: Seq<T>> nodup(s0) && s.exchange(s0, i, j) ==> nodup(s))]
fn swap_nodup<T>(s: Seq<T>, i: Int, j: Int) {
    proof_assert!{ forall<s0: Seq<T>> s.exchange(s0, i, j) ==> s.permutation_of(s0) };
    permut_nodup::<T>(Seq::EMPTY, Seq::EMPTY); // HACK
}

#[logic]
#[requires(1 <= i && i <= j && j <= s.len())]
#[ensures(forall<x:T> s[i-1] == x ==>
    Seq::ext_eq(s.subsequence(i, j).push_front(x), s.subsequence(i-1, j)))]
fn subsequence_extend_l<T>(s: Seq<T>, i: Int, j: Int) {}

// Knuth's Shuffle

#[requires(e.amount == Frac1(fact(v@.len())))]
#[requires(avoid.permutation_of(v@))]
#[requires(nodup(v@))]
#[ensures((^v)@.permutation_of(v@))]
#[ensures((^v)@ != *avoid)]
pub fn knuth_shuffle<T: Sized>(v: &mut Vec<T>, mut e: GhostBox<ErrCredit>, avoid: Snapshot<Seq<T>>) {
    let old_v = snapshot! { v };
    let n = snapshot!{ v@.len() };

    let mut different_at: Snapshot<Option<Int>> = snapshot!{ None };

    #[invariant((*v).invariant())]
    #[invariant(v@.permutation_of(old_v@))]
    #[invariant(avoid.permutation_of(old_v@))]
    #[invariant(nodup(v@))]
    #[invariant(*different_at == None ==>
       Seq::ext_eq(v@.subsequence(*n - produced.len(), *n), avoid.subsequence(*n - produced.len(), *n)) &&
       e.amount == Frac1(fact(*n - produced.len())))]
    #[invariant(forall<i:Int> *different_at == Some(i) ==>
       *n - produced.len() <= i && i < *n && v@[i] != (*avoid)[i])]
    // NOTE: the invariant holds (as stated) at the *end* of the loop body
    for i in 0..v.len() {
        let upper = v.len() - i;
        proof_assert!{ upper@ == *n - produced.len() + 1 };
        proof_assert!{ upper@ >= 1 };

        let avoid_val = snapshot!{ avoid[upper@ - 1] };
        let avoid_idx = snapshot!{ find_idx(v@, *avoid_val) };
        proof_assert!{ avoid.contains(*avoid_val) };
        proof_assert!{ v@.contains(*avoid_val) };
        proof_assert!{ *avoid_val == v@[*avoid_idx] };

        let (j, avoided) =
            rand_in_range(
                0,
                upper,
                e.borrow_mut(),
                avoid_idx,
            );

        proof_assert!{ *avoided && nodup(v@) ==> v@[j@] != *avoid_val };

        v.swap(j, upper - 1);
        proof_assert!{ *avoided ==> v@[upper@ - 1] != *avoid_val };
        proof_assert!{ !*avoided ==> v@[upper@ - 1] == *avoid_val };

        proof_assert!{ *different_at == None ==>
            Seq::ext_eq(v@.subsequence(upper@, *n),
                        avoid.subsequence(upper@, *n)) };
        // lemma function: a bit clunky
        proof_assert!{ swap_nodup(v@, j@, upper@ - 1); nodup(v@) };

        different_at = snapshot!{
            if *different_at == None && *avoided {
                Some(upper@ - 1)
            } else {
                *different_at
            }
        };

        proof_assert!{
            *different_at == None ==> {
            subsequence_extend_l(Seq::<T>::EMPTY, 0, 0); // HACK
            Seq::ext_eq(v@.subsequence(upper@ - 1, *n),
                        avoid.subsequence(upper@ - 1, *n)) }};

        proof_assert!{ *different_at == None ==>
            e.amount == Frac1(upper@ * fact(upper@ - 1)).mul_int(upper@) };
    }

    proof_assert!{ match *different_at {
        None => e.amount == Frac1(1) && e.impossible(),
        Some(_) => ! Seq::ext_eq(v@, *avoid),
    }};
}
```

**Comment by Lysxia** [comment]
Maybe solved by #1222

**Comment by Lysxia** [comment]
Fixed by #1222
